File: packages/go/sender_protocol_test.go
package crumbtrail

import (
	"context"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"
)

// The delivery contract: what the sender decides once it has an answer.
//
// TestSenderTLSRetryAndClose proves the HTTPS path works. This is the part where the damage
// was: a 202 shed and a 401 both read as success, so a revoked key produced a working looking
// SDK and a permanently empty project.

type deliveryLog struct {
	mu       sync.Mutex
	requests []Batch
}

func (d *deliveryLog) record(r *http.Request) {
	var batch Batch
	body, _ := io.ReadAll(io.LimitReader(r.Body, 1<<20))
	_ = json.Unmarshal(body, &batch)
	d.mu.Lock()
	defer d.mu.Unlock()
	d.requests = append(d.requests, batch)
}

func (d *deliveryLog) count() int {
	d.mu.Lock()
	defer d.mu.Unlock()
	return len(d.requests)
}

func (d *deliveryLog) gaps() []map[string]any {
	d.mu.Lock()
	defer d.mu.Unlock()
	var out []map[string]any
	for _, batch := range d.requests {
		for _, event := range batch.Events {
			if event.K == "capture_gap" {
				out = append(out, event.D)
			}
		}
	}
	return out
}

// answers replies with each status in turn, then 200 for anything after.
func newDelivery(t *testing.T, answer func(int, http.ResponseWriter)) (*Sender, *deliveryLog, *strings.Builder) {
	t.Helper()
	delivery := &deliveryLog{}
	server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		delivery.record(r)
		answer(delivery.count(), w)
	}))
	t.Cleanup(server.Close)
	messages := &strings.Builder{}
	sender, err := NewSender(SenderConfig{Endpoint: server.URL, Key: "test-key", HTTPClient: server.Client(),
		Logger: log.New(messages, "", 0)})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()
		_ = sender.Close(ctx)
	})
	return sender, delivery, messages
}

func batchOf(session string, count int) pendingBatch {
	events := make([]Event, count)
	for i := range events {
		events[i] = newEvent(int64(i), "backend.req.end", map[string]any{})
	}
	payload, _ := json.Marshal(Batch{session, events})
	return pendingBatch{session: session, payload: payload, events: count}
}

func TestPermanentStatusRecordsAGapAndLogs(t *testing.T) {
	sender, delivery, messages := newDelivery(t, func(n int, w http.ResponseWriter) {
		if n == 1 {
			w.WriteHeader(http.StatusUnauthorized)
			return
		}
		w.WriteHeader(http.StatusOK)
	})
	// deliver is driven directly so the assertions do not race the delivery goroutine.
	sender.deliver(batchOf("session", 4))
	if delivery.count() != 2 {
		t.Fatalf("a 401 was attempted %d times; the key is revoked, repeating cannot help", delivery.count()-1)
	}
	gaps := delivery.gaps()
	if len(gaps) != 1 {
		t.Fatal("a permanent refusal left no record that the evidence is gone")
	}
	if gaps[0]["reason"] != "delivery_failed" || gaps[0]["droppedEventCount"] != float64(4) || gaps[0]["detail"] != "HTTP 401" {
		t.Fatalf("gap %v", gaps[0])
	}
	if !strings.Contains(messages.String(), "HTTP 401") {
		t.Fatalf("nothing was logged: %q", messages.String())
	}
}

func TestNotFoundIsAttemptedOnce(t *testing.T) {
	sender, delivery, _ := newDelivery(t, func(n int, w http.ResponseWriter) {
		if n == 1 {
			w.WriteHeader(http.StatusNotFound)
			return
		}
		w.WriteHeader(http.StatusOK)
	})
	sender.deliver(batchOf("session", 1))
	if delivery.count() != 2 {
		t.Fatalf("%d requests; want the batch once plus its gap", delivery.count())
	}
}

func TestRetryableStatusIsRepeatedThenAccepted(t *testing.T) {
	sender, delivery, _ := newDelivery(t, func(n int, w http.ResponseWriter) {
		switch n {
		case 1:
			w.WriteHeader(http.StatusTooManyRequests)
		case 2:
			w.WriteHeader(http.StatusServiceUnavailable)
		default:
			w.WriteHeader(http.StatusOK)
		}
	})
	sender.deliver(batchOf("session", 2))
	if delivery.count() != 3 {
		t.Fatalf("%d attempts, want 3", delivery.count())
	}
	if gaps := delivery.gaps(); len(gaps) != 0 {
		t.Fatalf("a batch that was eventually accepted reported a hole: %v", gaps)
	}
}

// A 202 passes every "is this a success" test while the cloud has already discarded the batch.
func TestShedPausesDeliveryAndReportsTheWholeHole(t *testing.T) {
	sender, delivery, messages := newDelivery(t, func(n int, w http.ResponseWriter) {
		if n == 1 {
			w.WriteHeader(http.StatusAccepted)
			io.WriteString(w, `{"capture":"shed","reason":"rate_limited_ingest","retryAfterSeconds":0.05}`)
			return
		}
		w.WriteHeader(http.StatusOK)
	})
	sender.deliver(batchOf("session", 3))
	if delivery.count() != 1 {
		t.Fatal(delivery.count())
	}
	if !strings.Contains(messages.String(), "rate_limited_ingest") {
		t.Fatalf("the shed window was not logged: %q", messages.String())
	}
	sender.deliver(batchOf("session", 5))
	if delivery.count() != 1 {
		t.Fatal("delivery continued inside the Retry-After window")
	}
	time.Sleep(80 * time.Millisecond)
	sender.deliver(batchOf("session", 1))
	if delivery.count() != 3 {
		t.Fatalf("%d requests; delivery did not resume with its held gap after the window", delivery.count())
	}
	gaps := delivery.gaps()
	if len(gaps) != 1 || gaps[0]["reason"] != "rate_limited_ingest" {
		t.Fatalf("gaps %v", gaps)
	}
	if gaps[0]["droppedEventCount"] != float64(8) {
		t.Fatalf("dropped %v; the shed batch and the suppressed batch are both lost", gaps[0]["droppedEventCount"])
	}
}

func TestRetryAfterHeaderIsHonouredWithoutABodyField(t *testing.T) {
	sender, delivery, _ := newDelivery(t, func(n int, w http.ResponseWriter) {
		if n == 1 {
			w.Header().Set("Retry-After", "120")
			w.WriteHeader(http.StatusAccepted)
			io.WriteString(w, `{"capture":"shed","reason":"kill_switch"}`)
			return
		}
		w.WriteHeader(http.StatusOK)
	})
	sender.deliver(batchOf("session", 1))
	sender.deliver(batchOf("session", 1))
	if delivery.count() != 1 {
		t.Fatal("the Retry-After header was ignored")
	}
}

// An unrecognised reason is a classification this SDK cannot vouch for.
func TestUnknownShedReasonIsRecordedAsAPlainDeliveryFailure(t *testing.T) {
	sender, delivery, _ := newDelivery(t, func(n int, w http.ResponseWriter) {
		if n == 1 {
			w.WriteHeader(http.StatusAccepted)
			io.WriteString(w, `{"capture":"shed","reason":"something_new","retryAfterSeconds":0}`)
			return
		}
		w.WriteHeader(http.StatusOK)
	})
	sender.deliver(batchOf("session", 2))
	sender.deliver(batchOf("session", 1))
	gaps := delivery.gaps()
	if len(gaps) != 1 || gaps[0]["reason"] != "delivery_failed" {
		t.Fatalf("gaps %v", gaps)
	}
}

// Without this the first refusal posts a gap, the gap is refused, and that posts a gap.
func TestARefusedGapDoesNotRecurse(t *testing.T) {
	sender, delivery, _ := newDelivery(t, func(_ int, w http.ResponseWriter) { w.WriteHeader(http.StatusUnauthorized) })
	sender.deliver(batchOf("session", 1))
	if delivery.count() != 2 {
		t.Fatalf("%d requests, want the batch and one gap", delivery.count())
	}
}

func TestCloseDoesNotWaitOutTheRetryBudget(t *testing.T) {
	sender, _, _ := newDelivery(t, func(_ int, w http.ResponseWriter) { w.WriteHeader(http.StatusServiceUnavailable) })
	sender.Enqueue(Batch{"session", []Event{newEvent(1, "backend.req.end", map[string]any{})}})
	started := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
	defer cancel()
	_ = sender.Close(ctx)
	if elapsed := time.Since(started); elapsed > time.Second {
		t.Fatalf("close waited %s for the retry backoff", elapsed)
	}
}

File: packages/ruby/test/sender_protocol_test.rb
require 'minitest/autorun'
require 'json'
require 'crumbtrail'

# The delivery contract, driven through a stubbed transport rather than a socket.
#
# `test_real_tls_delivery_retries_and_drains` in capture_test.rb proves the real HTTPS path
# works. These tests prove what the sender decides once it has an answer, which is where the
# damage was: a 202 shed and a 401 both read as success, so a revoked key produced a working
# looking SDK and a permanently empty project.
class SenderProtocolTest < Minitest::Test
  Response = Struct.new(:code, :body, :headers) do
    def [](name) = (headers || {})[name]
  end

  class Logger
    attr_reader :messages
    def initialize = @messages = []
    def warn(message) = @messages << message
  end

  def build(*responses)
    logger = Logger.new
    sender = Crumbtrail::Sender.new(endpoint: 'https://ingest.example.com', key: 'test-key', logger: logger)
    posts = []
    queued = responses.dup
    sender.define_singleton_method(:post) do |payload|
      posts << JSON.parse(payload)
      queued.shift || Response.new('200', '', {})
    end
    [sender, posts, logger]
  end

  def batch(session, count)
    events = Array.new(count) { |index| Crumbtrail.event(index, 'backend.req.end', {}) }
    JSON.generate(sessionId: session, events: events)
  end

  # `deliver` is driven directly so the assertions do not race the worker thread.
  def deliver(sender, session, count)
    sender.send(:deliver, session, batch(session, count), count)
  end

  def gaps(posts)
    posts.flat_map { |post| post['events'] }.select { |event| event['k'] == 'capture_gap' }
  end

  def test_permanent_status_records_a_gap_and_logs
    sender, posts, logger = build(Response.new('401', '', {}))
    deliver(sender, 'session', 4)
    assert_equal 2, posts.size, 'a 401 was retried; the key is revoked, repeating cannot help'
    gap = gaps(posts).first
    refute_nil gap, 'a permanent refusal left no record that the evidence is gone'
    assert_equal 'delivery_failed', gap['d']['reason']
    assert_equal 4, gap['d']['droppedEventCount']
    assert_equal 'HTTP 401', gap['d']['detail']
    assert_includes logger.messages.first, 'HTTP 401'
  ensure
    sender&.close(timeout: 1)
  end

  def test_not_found_is_attempted_once
    sender, posts, = build(Response.new('404', '', {}))
    deliver(sender, 'session', 1)
    assert_equal 1, posts.count { |post| post['events'].none? { |event| event['k'] == 'capture_gap' } }
  ensure
    sender&.close(timeout: 1)
  end

  def test_retryable_status_is_repeated_then_accepted
    sender, posts, = build(Response.new('429', '', {}), Response.new('503', '', {}), Response.new('200', '', {}))
    deliver(sender, 'session', 2)
    assert_equal 3, posts.size
    assert_empty gaps(posts), 'a batch that was eventually accepted reported a hole'
  ensure
    sender&.close(timeout: 1)
  end

  # A 202 passes every "is this a success" test while the cloud has already discarded the batch.
  def test_shed_pauses_delivery_and_reports_the_whole_hole
    shed = Response.new('202', JSON.generate(capture: 'shed', reason: 'rate_limited_ingest', retryAfterSeconds: 0.05), {})
    sender, posts, logger = build(shed)
    deliver(sender, 'session', 3)
    assert_equal 1, posts.size
    assert_includes logger.messages.first, 'rate_limited_ingest'
    deliver(sender, 'session', 5)
    assert_equal 1, posts.size, 'delivery continued inside the Retry-After window'
    sleep 0.08
    deliver(sender, 'session', 1)
    assert_equal 3, posts.size, 'delivery did not resume after the Retry-After window'
    gap = gaps(posts).first
    assert_equal 'rate_limited_ingest', gap['d']['reason']
    assert_equal 8, gap['d']['droppedEventCount'], 'the shed batch and the suppressed batch are both lost'
  ensure
    sender&.close(timeout: 1)
  end

  def test_retry_after_header_is_honoured_without_a_body_field
    shed = Response.new('202', JSON.generate(capture: 'shed', reason: 'kill_switch'), { 'Retry-After' => '120' })
    sender, posts, = build(shed)
    deliver(sender, 'session', 1)
    deliver(sender, 'session', 1)
    assert_equal 1, posts.size, 'the Retry-After header was ignored'
  ensure
    sender&.close(timeout: 1)
  end

  # An unrecognised reason is a classification this SDK cannot vouch for.
  def test_unknown_shed_reason_is_recorded_as_a_plain_delivery_failure
    shed = Response.new('202', JSON.generate(capture: 'shed', reason: 'something_new', retryAfterSeconds: 0), {})
    sender, posts, = build(shed)
    deliver(sender, 'session', 2)
    deliver(sender, 'session', 1)
    assert_equal 'delivery_failed', gaps(posts).first['d']['reason']
  ensure
    sender&.close(timeout: 1)
  end

  # Without this the first refusal posts a gap, the gap is refused, and that posts a gap.
  def test_a_refused_gap_does_not_recurse
    sender, posts, = build(*Array.new(6) { Response.new('401', '', {}) })
    deliver(sender, 'session', 1)
    assert_equal 2, posts.size
  ensure
    sender&.close(timeout: 1)
  end

  def test_close_does_not_wait_out_the_retry_budget
    sender, = build(*Array.new(4) { Response.new('503', '', {}) })
    sender.enqueue(sessionId: 'session', events: [Crumbtrail.event(1, 'backend.req.end', {})])
    started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    sender.close(timeout: 0.1)
    assert_operator Process.clock_gettime(Process::CLOCK_MONOTONIC) - started, :<, 1.2
  end
end
